learner/sampler/slurper.py

```diff
@@ -0,0 +1,252 @@
+                        line_length = len(line)
+                        f.seek(0,2)
+                        eof = f.tell()
+                        average_length = 0
+                        for sample in range(2,32):
+                            f.seek(math.floor(nlines/sample),0)
+                            f.readline()
+                            line = f.readline()
+                            average_length += len(line)
+                        f.close()
+                    return math.floor(eof/(average_length/30))
+                except:
+                    continue
+            print("Cannot estimate line size using the typical encodings. Is your data semi-structured?")
+            
+        return math.floor(eof/(average_length/30))
+            
+    def get_sample_size(nlines: int = None, confidence: float = 0.95, error: float = 0.05) -> int:
+        """Returns number of samples needed for a population of nlines, when sampling is done with replacement"""
+        if not nlines:
+            nlines = self.nlines
+            
+        #Hard coding the critical and std. dev. for now
+        X = (1.96 * 0.5 * 0.5)/(error**2)
+        return math.floor(nlines*X / (X + N - 1))
+            
+    
+    def pythonic_reservoir(file: str = None, encoding: str = None, reservoir_size: int = None) -> List[str]:
+        """Make a single pass through the file, replacing each value with some probability. This is Knuth's Reservoir Sampling."""
+        if not file:
+            file = self.filename
+        if not reservoir_size:
+            reservoir_size = self.sample_size
+        if not encoding:
+            encoding = self.encoding              
+            
+        reservoir = []
+        n = reservoir_size - 1
+        counter = 0
+        try:
+            with open(file, encoding = encoding) as f:
+                for line in f:
+                    if counter < reservoir_size:
+                        reservoir.append(line.strip())
+                    else:
+                        n += 1
+                        draw = random.randrange(0,n,1)
+                        if draw < reservoir_size:
+                            reservoir[draw] = line
+                f.close()
+        except:
+            for cycle in range(1,7):
+                new_encoding = self.check_file_encodings(file = file, encoding = None, cycle_bit = cycle)
+                try:
+                    with open(file, encoding = new_encoding) as f:
+                        for line in f:
+                            if counter < reservoir_size:
+                                reservoir.append(line.strip())
+                            else:
+                                n += 1
+                                draw = random.randrange(0,n,1)
+                                if draw < reservoir_size:
+                                    reservoir[draw] = line
+                        f.close()
+                    return reservoir
```

Comment:
I feel as though again, this could easily be broken down into multiple functions. Also we should funnel into a single point early on where we figure out what the encoding is, and then it is a safe assumption that the encoding won't change after that. Feels like you expect these functions to be called out of turn. Don't. don't provide the option for a Optional encoding.